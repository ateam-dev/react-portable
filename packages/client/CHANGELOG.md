# @react-portable/client

## 0.0.2

### Patch Changes

- [#3](https://github.com/ateam-dev/react-portable/pull/3) [`d073b81`](https://github.com/ateam-dev/react-portable/commit/d073b8196680285c85bf70a5484df4fda7c8dff3) Thanks [@aiji42](https://github.com/aiji42)! - first release!

## 0.0.1

### Patch Changes

- [#1](https://github.com/ateam-dev/react-portable/pull/1) [`9e7866c`](https://github.com/ateam-dev/react-portable/commit/9e7866c26b7ca994a5d32a177759f66bafe3800c) Thanks [@aiji42](https://github.com/aiji42)! - first release
