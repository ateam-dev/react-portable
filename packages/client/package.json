{
  "name": "@react-portable/client",
  "version": "0.5.0",
  "repository": "git@github.com:ateam-dev/react-portable.git",
  "license": "Apache-2.0",
  "main": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
      "default": "./dist/index.mjs",
      "types": "./dist/index.d.ts"
    },
    "./web-components": {
      "import": "./dist/web-components.mjs",
      "require": "./dist/web-components.cjs",
      "default": "./dist/web-components.mjs",
      "types": "./dist/web-components.d.ts"
    },
    "./browser": {
      "default": "./dist/browser.umd.js"
    },
    "./*": "./*"
  },
  "typesVersions": {
    "*": {
      "web-components": [
        "./dist/web-components.d.ts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build:main": "unbuild",
    "build:browser": "vite build",
    "build": "run-s build:browser build:main",
    "watch:main": "nodemon --ext ts,tsx --watch src -x 'npm run build:main'",
    "watch:browser": "vite build --watch",
    "watch": "run-p watch:*",
    "preview": "vite preview --port 3003",
    "dev": "run-p watch preview",
    "test": "vitest run",
    "test:coverage": "vitest run --coverage",
    "typecheck": "tsc --noEmit"
  },
  "devDependencies": {
    "@vitest/coverage-v8": "^0.33.0",
    "jsdom": "^22.0.0",
    "msw": "^1.2.2",
    "nodemon": "^3.0.1",
    "npm-run-all": "^4.1.5",
    "unbuild": "^2.0.0",
    "vite": "^4.4.3",
    "vite-plugin-dts": "^3.2.0",
    "vitest": "^0.33.0"
  },
  "dependencies": {
    "writable-dom": "^1.0.3"
  },
  "peerDependencies": {
    "react": ">=18"
  },
  "unbuild": {
    "clean": false
  }
}
