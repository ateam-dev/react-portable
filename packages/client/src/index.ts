export {
  registerReactPortable,
  parseFragmentId,
  createFragmentId,
  CLASS_NAME_FOR_GATEWAY_CACHE,
  CUSTOM_HEADER_KEY_GATEWAY,
} from "./react-portable";
