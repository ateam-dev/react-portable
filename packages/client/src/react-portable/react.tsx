import { useEffect, useReducer, useRef, useId } from "react";
import { register } from "./register";

export const ReactPortable = ({
  entry,
  gateway,
}: {
  entry: string;
  gateway?: string;
}) => {
  const isMounted = useMounted();
  const id = useId();
  const ref = useRef("");
  if (!isMounted && typeof window !== "undefined") {
    ref.current = document.getElementById(id)?.innerText ?? "";
    register();
  }

  return (
    <react-portable
      gateway={gateway}
      entry={entry}
      id={id}
      dangerouslySetInnerHTML={{ __html: ref.current }}
    />
  );
};

const useMounted = () => {
  const [isMounted, mounted] = useReducer(() => true, false);
  useEffect(() => {
    mounted();
  }, []);

  return isMounted;
};
