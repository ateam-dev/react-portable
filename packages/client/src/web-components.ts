export { register } from "./rp-preview/register";
export { RpPreview } from "./rp-preview/element";
