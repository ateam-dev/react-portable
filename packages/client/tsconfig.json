{
  "compilerOptions": {
    "target": "ESNext",
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "skipLibCheck": true,
    "strict": true,
    "module": "ESNext",
    "moduleResolution": "Node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "allowSyntheticDefaultImports": true
  },
  "include": ["src"],
}
