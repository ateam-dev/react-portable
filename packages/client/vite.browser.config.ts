import { defineConfig } from "vite";

export default defineConfig({
  build: {
    lib: {
      entry: "src/browser.ts",
      formats: ["umd"],
      name: "ReactPortable",
      fileName: "browser",
    },
  },
});
