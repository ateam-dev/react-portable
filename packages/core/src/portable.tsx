import React, { FunctionComponent, useEffect, useRef } from "react";
import { RpPreview } from "@react-portable/client/web-components";
import { RequestEventCommon } from "@builder.io/qwik-city/middleware/request-handler";

export type Strategy = {
  revalidate?: false | 0 | number;
  hydrate?: "onUse" | "onIdle" | "disable";
};

type ErrorResponse = ReturnType<RequestEventCommon["error"]>;

export type Loader<T extends Record<string, unknown> = {}> = (
  r: Request,
  ctx: { error: RequestEventCommon["error"] },
) => T | ErrorResponse | Promise<T | ErrorResponse>;

export interface PortableComponent<T extends Record<string, unknown> = {}>
  extends FunctionComponent<T> {
  __code: string;
  __strategy: undefined | Strategy;
  __loader: undefined | Loader<T>;
  __original: FunctionComponent<T>;
}

type InferProps<T> = T extends FunctionComponent<infer U>
  ? U
  : T extends (props: infer U) => JSX.Element
  ? U
  : T extends () => JSX.Element
  ? {}
  : never;

type PortableOption<
  T extends FunctionComponent | ((props: any) => JSX.Element),
> = {
  loader?: Loader<InferProps<T>>;
  strategy?: Strategy;
  disablePreview?: boolean;
};

export const portable = <
  T extends FunctionComponent | ((props: any) => JSX.Element),
>(
  Component: T,
  code: string,
  { loader, strategy, disablePreview = false }: PortableOption<T> = {},
): PortableComponent<InferProps<T>> => {
  const Wrapped = (props: InferProps<T>) => {
    const ref = useRef<RpPreview>(null);
    useEffect(() => {
      if (!ref.current) return;

      ref.current.props = props;
      ref.current.rerender?.();
    }, [props]);

    if (disablePreview) return <Component {...(props as any)} />;

    return (
      <rp-preview ref={ref} code={code}>
        <Component
          {...(props as any)}
          children={<rp-outlet>{props.children}</rp-outlet>}
        />
      </rp-preview>
    );
  };

  Wrapped.__code = code;
  Wrapped.__loader = loader;
  Wrapped.__strategy = strategy;
  Wrapped.__original = Component;

  return Wrapped;
};

export const previewify = <
  T extends FunctionComponent | ((props: any) => JSX.Element),
>(
  Component: T,
  code: string,
): PortableComponent<InferProps<T>> => {
  return portable(Component, code, { strategy: { hydrate: "onIdle" } });
};
