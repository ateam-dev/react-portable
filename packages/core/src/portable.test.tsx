/**
 * @vitest-environment jsdom
 */
import { beforeEach, describe, expect, test, vi } from "vitest";
import { portable } from "./portable";
import { render, cleanup, act, fireEvent } from "@testing-library/react";
import { rpPreviewRegister } from "@react-portable/client/web-components";
import { MouseEventHandler, ReactNode } from "react";

const Sample = ({
  onClick,
  testId,
}: {
  onClick?: MouseEventHandler<HTMLDivElement>;
  testId?: string;
}) => (
  <div onClick={onClick} data-testid={testId}>
    sample
  </div>
);

const SampleWithChildren = ({
  children,
  element,
}: {
  children: ReactNode;
  element?: ReactNode;
}) => (
  <div>
    sample with children:{children} {element}
  </div>
);

describe("portable", () => {
  beforeEach(() => {
    vi.spyOn(crypto, "randomUUID").mockReturnValueOnce(
      "dummy-uuid" as ReturnType<typeof crypto.randomUUID>,
    );
    cleanup();
    rpPreviewRegister();
  });

  test("__code", () => {
    const Component = portable(Sample, "foo");
    expect(Component.__code).toBe("foo");
  });
  test("__strategy", () => {
    const Component = portable(Sample, "foo", {
      strategy: { hydrate: "onUse", revalidate: 60 },
    });
    expect(Component.__strategy).toStrictEqual({
      hydrate: "onUse",
      revalidate: 60,
    });
  });
  test("__loader", () => {
    const loader = () => {
      console.log("loader");
      return {};
    };
    const Component = portable(Sample, "foo", {
      loader,
    });
    expect(Component.__loader).toBe(loader);
  });

  describe("Component", () => {
    test("It is same as the original component when initial", () => {
      const Component = portable(Sample, "foo");
      const wrapped = render(<Component />);
      const original = render(<Sample />);

      expect(wrapped.container.innerHTML).toBe(original.container.innerHTML);
    });

    test("Dispatchers are added to rpPreviewDispatchers when rendered and removed when unmounted", () => {
      const Component = portable(Sample, "foo");
      const { unmount } = render(<Component />);
      render(<Component />);

      expect(window.rpPreviewDispatchers?.length).toBe(2);

      unmount();

      expect(window.rpPreviewDispatchers?.length).toBe(1);
    });

    test("It will be wrapped by <rp-preview /> when rpPreviewDispatchers will be called", () => {
      const Component = portable(Sample, "foo");
      const { asFragment } = render(<Component />);

      act(() =>
        window.rpPreviewDispatchers!.forEach(([, dispatch]) => dispatch()),
      );

      expect(asFragment()).toMatchSnapshot();
    });

    test("Elements of type ReactElement will be also rendered in the template when rpPreviewDispatchers will be called", async () => {
      const Component = portable(SampleWithChildren, "foo");
      const { container } = render(
        <Component
          children={<div>children</div>}
          element={<div>element</div>}
        />,
      );

      act(() =>
        window.rpPreviewDispatchers!.forEach(([, dispatch]) => dispatch()),
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe("Component.__forQwik", () => {
    test("`__outlet__` is converted to <rp-outlet /> and rendered", () => {
      const Component = portable(SampleWithChildren, "foo");
      const { asFragment } = render(
        <Component.__forQwik children="__outlet__" element="__outlet__" />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
    test("`__function__` is converted into a function that delegates the original handler", () => {
      const onClickSpy = vi.fn();
      const Component = portable(Sample, "foo");
      const rendered = render(<Component onClick={onClickSpy} />);
      render(
        <Component.__forQwik
          // @ts-ignore
          onClick="__function__#dummy-uuid#rp-preview-event"
          testId="clickable"
        />,
      );

      act(() =>
        window.rpPreviewDispatchers!.forEach(([, dispatch]) => dispatch()),
      );

      fireEvent.click(rendered.getByTestId("clickable"));

      expect(onClickSpy).toBeCalled();
    });
  });
});
