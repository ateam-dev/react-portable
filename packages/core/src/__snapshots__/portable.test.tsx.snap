// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`portable > hoge 1`] = `
<DocumentFragment>
  <react-portable-preview
    code="foo"
  >
    <div>
      sample foo
    </div>
  </react-portable-preview>
</DocumentFragment>
`;

exports[`portable > the component will NOT be wrapped by react-portable-preview when \`disablePreview\` is specified true 1`] = `
<DocumentFragment>
  <div>
    sample
  </div>
</DocumentFragment>
`;

exports[`portable > the component will be wrapped by react-portable-preview 1`] = `
<DocumentFragment>
  <react-portable-preview
    code="foo"
  >
    <div>
      sample
    </div>
  </react-portable-preview>
</DocumentFragment>
`;
